client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/operation/CompareAndSetOperation.java

```diff
@@ -0,0 +1,58 @@
+package org.infinispan.client.hotrod.counter.operation;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.infinispan.client.hotrod.configuration.Configuration;
+import org.infinispan.client.hotrod.impl.protocol.Codec;
+import org.infinispan.client.hotrod.impl.protocol.HeaderParams;
+import org.infinispan.client.hotrod.impl.transport.Transport;
+import org.infinispan.client.hotrod.impl.transport.TransportFactory;
+import org.infinispan.counter.api.CounterConfiguration;
+import org.infinispan.counter.api.StrongCounter;
+import org.infinispan.counter.exception.CounterOutOfBoundsException;
+
+/**
+ * A compare-and-set operation for {@link StrongCounter#compareAndSet(long, long)}.
+ *
+ * @author Pedro Ruivo
+ * @since 9.2
+ */
+public class CompareAndSetOperation extends BaseCounterOperation<Boolean> {
+
+   private final long expect;
+   private final long update;
+   private final CounterConfiguration counterConfiguration;
+
+   public CompareAndSetOperation(Codec codec, TransportFactory transportFactory, AtomicInteger topologyId,
+         Configuration cfg, String counterName, long expect, long update,
+         CounterConfiguration counterConfiguration) {
+      super(codec, transportFactory, topologyId, cfg, counterName);
+      this.expect = expect;
+      this.update = update;
+      this.counterConfiguration = counterConfiguration;
+   }
+
+   @Override
+   protected Boolean executeOperation(Transport transport) {
+      HeaderParams header = writeHeaderAndCounterName(transport, COUNTER_CAS_REQUEST);
+      transport.writeLong(expect);
+      transport.writeLong(update);
+      transport.flush();
+
+      short status = readHeaderAndValidateCounter(transport, header);
+      assertBoundaries(status);
+      return status == NO_ERROR_STATUS;
+   }
+
+   private void assertBoundaries(short status) {
+      if (status == NOT_EXECUTED_WITH_PREVIOUS) {
+         if (update >= counterConfiguration.upperBound()) {
+            throw new CounterOutOfBoundsException(
```

Comment:
it is correct. I use the `NOT_EXECUTED_WITH_PREVIOUS` to signal counter out of bound exception. I just identifying if it was the lower or upper bound to avoid return more data in the network.

Reply:
My bad.